File: ios/BiblosSurferTests/Helpers/LocatorRoundTripTests.swift
//
//  LocatorRoundTripTests.swift
//  BiblosSurferTests
//

import ReadiumShared
import XCTest
@testable import BiblosSurfer

/// The reading position is stored as `Locator` JSON, so a broken round trip means every reader
/// silently reopens at page one. These tests also pin the wire format, which has to stay readable by
/// the Kotlin toolkit if the Android port ever lands.
final class LocatorRoundTripTests: XCTestCase {
    func testLocatorSurvivesJSONRoundTrip() throws {
        let original = TestFixtures.locator()

        let json = try original.jsonString()
        let restored = try Locator(jsonString: json)

        XCTAssertEqual(restored.href, original.href)
        XCTAssertEqual(restored.mediaType, original.mediaType)
        XCTAssertEqual(restored.title, original.title)
        XCTAssertEqual(restored.locations.progression, original.locations.progression)
        XCTAssertEqual(restored.locations.totalProgression, original.locations.totalProgression)
        XCTAssertEqual(restored.locations.position, original.locations.position)
        XCTAssertEqual(restored.text.highlight, original.text.highlight)
    }

    func testJSONUsesTheKeysSharedWithTheKotlinToolkit() throws {
        let json = try TestFixtures.locator().jsonString()
        let object = try XCTUnwrap(
            JSONSerialization.jsonObject(with: XCTUnwrap(json.data(using: .utf8))) as? [String: Any]
        )

        XCTAssertNotNil(object["href"])
        XCTAssertEqual(object["type"] as? String, MediaType.xhtml.string)
        XCTAssertNotNil(object["locations"])
        XCTAssertNotNil(object["text"])
    }

    func testMinimalLocatorRoundTripsWithoutOptionalFields() throws {
        let original = TestFixtures.locator(
            title: nil,
            progression: nil,
            totalProgression: nil,
            position: nil,
            highlight: nil
        )

        let restored = try Locator(jsonString: try original.jsonString())

        XCTAssertEqual(restored.href, original.href)
        XCTAssertNil(restored.title)
        XCTAssertNil(restored.locations.progression)
        XCTAssertNil(restored.locations.position)
    }
}

File: android/app/src/test/java/miko/biblossurfer/helpers/LocatorRoundTripTests.kt
package miko.biblossurfer.helpers

import androidx.test.ext.junit.runners.AndroidJUnit4
import miko.biblossurfer.support.TestFixtures
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.util.mediatype.MediaType
import org.robolectric.annotation.Config

/**
 * The reading position is stored as Locator JSON, so a broken round trip means every reader
 * silently reopens at page one. These tests also pin the wire format, which has to stay readable by
 * the Swift toolkit.
 */
@RunWith(AndroidJUnit4::class)
@Config(sdk = [33])
class LocatorRoundTripTests {
    @Test
    fun locatorSurvivesJSONRoundTrip() {
        val original = TestFixtures.locator()
        val json = original.toJSON()
        val restored = Locator.fromJSON(json)!!

        assertEquals(original.href, restored.href)
        assertEquals(original.mediaType, restored.mediaType)
        assertEquals(original.title, restored.title)
        assertEquals(original.locations.progression, restored.locations.progression)
        assertEquals(original.locations.totalProgression, restored.locations.totalProgression)
        assertEquals(original.locations.position, restored.locations.position)
        assertEquals(original.text.highlight, restored.text.highlight)
    }

    @Test
    fun jsonUsesTheKeysSharedWithTheSwiftToolkit() {
        val json = TestFixtures.locator().toJSON()
        assertNotNull(json.opt("href"))
        assertEquals(MediaType.XHTML.toString(), json.optString("type"))
        assertNotNull(json.optJSONObject("locations"))
        assertNotNull(json.optJSONObject("text"))
    }

    @Test
    fun minimalLocatorRoundTripsWithoutOptionalFields() {
        val original = TestFixtures.locator(
            title = null,
            progression = null,
            totalProgression = null,
            position = null,
            highlight = null,
        )
        val restored = Locator.fromJSON(original.toJSON())!!
        assertEquals(original.href, restored.href)
        assertNull(restored.title)
        assertNull(restored.locations.progression)
        assertNull(restored.locations.position)
    }
}
